common/crypto/platform/openssl/crypto.cpp

```diff
@@ -231,6 +260,76 @@ expected::ExpectedString Sign(string private_key_path, vector<uint8_t> raw_data)
 	return EncodeBase64(signature);
 }
 
+expected::ExpectedBool VerifySignData(
+	const string public_key_path, mender::sha::SHA &shasum, vector<uint8_t> &signature) {
+	// load key
+	auto bio_key =
+		unique_ptr<BIO, void (*)(BIO *)>(BIO_new_file(public_key_path.c_str(), "r"), bio_free_func);
+	if (!bio_key.get()) {
+		return expected::unexpected(
+			MakeError(SetupError, "Failed to open the public key file" + GetOpenSSLErrorMessage()));
+	}
+	auto pkey = unique_ptr<EVP_PKEY, void (*)(EVP_PKEY *)>(
+		PEM_read_bio_PUBKEY(bio_key.get(), nullptr, nullptr, nullptr), pkey_free_func);
+	if (pkey == nullptr) {
+		return expected::unexpected(
+			MakeError(SetupError, "Failed to load the key" + GetOpenSSLErrorMessage()));
+	}
+
+	// prepare context
+	auto pkey_signer_ctx = unique_ptr<EVP_PKEY_CTX, void (*)(EVP_PKEY_CTX *)>(
+		EVP_PKEY_CTX_new(pkey.get(), nullptr), pkey_ctx_free_func);
+
+	auto ret = EVP_PKEY_verify_init(pkey_signer_ctx.get());
+	if (ret <= 0) {
+		return expected::unexpected(MakeError(
+			SetupError, "Failed to initialize the OpenSSL signer " + GetOpenSSLErrorMessage()));
+	}
+	ret = EVP_PKEY_CTX_set_rsa_padding(pkey_signer_ctx.get(), RSA_PKCS1_PADDING);
+	if (ret <= 0) {
+		return expected::unexpected(MakeError(
+			SetupError,
+			"Failed to set the OpenSSL padding to RSA_PKCS1" + GetOpenSSLErrorMessage()));
+	}
+	ret = EVP_PKEY_CTX_set_signature_md(pkey_signer_ctx.get(), EVP_sha256());
+	if (ret <= 0) {
+		return expected::unexpected(MakeError(
+			SetupError,
+			"Failed to set the OpenSSL signature to sha256" + GetOpenSSLErrorMessage()));
+	}
+
+	// verify signature
+	ret = EVP_PKEY_verify(
+		pkey_signer_ctx.get(), signature.data(), signature.size(), shasum.data(), shasum.size());
+	if (ret <= 0) {
```

Comment:
`0` is returned if the verification fails not because of an error, but because the signature is not valid or there was some mismatch. So this should be `< 0`, I believe.

Reply:
Yes, you are completely correct. On invalid signature we should not return an error, but just `false`. I'll fix